kubeapiversion/api_version.go

```diff
@@ -0,0 +1,172 @@
+package kubeapiversion
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/solo-io/go-utils/errors"
+)
+
+type PrereleaseModifier int
+
+var (
+	apiVersionRegexp *regexp.Regexp
+
+	MalformedVersionError = func(version string) error {
+		return errors.Errorf("Failed to parse kubernetes api version from %v", version)
+	}
+
+	InvalidMajorVersionError = errors.Errorf("Major version cannot be zero")
+
+	InvalidPrereleaseVersionError = errors.Errorf("Prerelease version cannot be zero")
+)
+
+const (
+	Alpha PrereleaseModifier = iota + 1
+	Beta
+	GA
+)
+
+func init() {
+	apiVersionRegexp = regexp.MustCompile(`^v([0-9]+)((alpha|beta)([0-9]+))?$`)
+}
+
+type ApiVersion interface {
+	Major() int
+	Prerelease() int
+	PrereleaseModifier() PrereleaseModifier
+	String() string
+	GreaterThan(other ApiVersion) bool
+	LessThan(other ApiVersion) bool
+	Equal(other ApiVersion) bool
+}
+
+type apiVersion struct {
+	major, prerelease int
+	modifier          PrereleaseModifier
+}
+
+func ParseApiVersion(version string) (ApiVersion, error) {
+	matches := apiVersionRegexp.FindStringSubmatch(version)
+	if matches == nil {
+		return nil, MalformedVersionError(version)
+	}
+
+	major, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return nil, err
+	}
+	if major == 0 {
+		return nil, InvalidMajorVersionError
+	}
+
+	// Prerelease version info is optional
+	var prerelease int
+	if matches[3] != "" && matches[4] != "" {
+		prerelease, err = strconv.Atoi(matches[4])
+		if err != nil {
+			return nil, err
+		}
+		if prerelease == 0 {
+			return nil, InvalidPrereleaseVersionError
+		}
+	}
+
+	var modifier PrereleaseModifier
+	switch matches[3] {
+	case "alpha":
+		modifier = Alpha
+	case "beta":
+		modifier = Beta
+	default:
+		modifier = GA
+	}
+
+	return &apiVersion{
+		major:      major,
+		prerelease: prerelease,
+		modifier:   modifier,
+	}, nil
+}
+
+func NewApiVersion(major, prerelease int, prereleaseModifier PrereleaseModifier) ApiVersion {
+	return &apiVersion{
+		major:      major,
+		prerelease: prerelease,
+		modifier:   prereleaseModifier,
+	}
+}
+
+func (v *apiVersion) Major() int {
+	return v.major
+}
+
+func (v *apiVersion) Prerelease() int {
+	return v.prerelease
+}
+
+func (v *apiVersion) PrereleaseModifier() PrereleaseModifier {
```

Comment:
I think PreReleaseModifier should probably have a `.String()` method to make it easier to consume